### PR TITLE
AzureRm.Storage module consuming api-version 2016-01-01 and common runtime

### PR DESCRIPTION
**Pull request by bganapa**
## Description
AzureRM.Storage module version update to 2016-01-01
Copying Storage SDK version 5.0.0-preview
Corresponding cmdlet changes, redoing IStorageContext move and Common runtime move, Redoing HashTable[] to HashTable
references to 8.6.0 WindowsAzure.Storage.dll and dependent dlls update
Recorded test changes
## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/StackAdmin/Storage.Management/AzureRM.Storage.psd1 [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+DotNetFrameworkVersion = '4.5.2'
+
+# Minimum version of the common language runtime (CLR) required by this module. This prerequisite is valid for the PowerShell Desktop edition only.
+CLRVersion = '4.0'
+
+# Processor architecture (None, X86, Amd64) required by this module
+# ProcessorArchitecture = ''
+
+# Modules that must be imported into the global environment prior to importing this module
+RequiredModules = @(@{ModuleName = 'AzureRM.Profile'; ModuleVersion = '5.3.4'; },
```
Comment (bganapa) [inline src/StackAdmin/Storage.Management/AzureRM.Storage.psd1]: Update AzureRm.Profile version before release

File: src/StackAdmin/Storage.Management/AzureRM.Storage.psd1 [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+               @{ModuleName = 'Azure.Storage'; ModuleVersion = '4.1.1'; })
```
Comment (bganapa) [inline src/StackAdmin/Storage.Management/AzureRM.Storage.psd1]: Azure.Storage version is already published

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj [hunk 3]
```diff
@@ -146,7 +125,30 @@
     <None Include="packages.config">
       <SubType>Designer</SubType>
     </None>
-    <None Include="StartupScripts\*.ps1" />
+    <None Include="StartupScripts\*.ps1">
+    </None>
+  </ItemGroup>
+  <ItemGroup>
+    <!--    <ProjectReference Include="..\..\..\Common\Commands.Common.Authentication.Abstractions\Commands.Common.Authentication.Abstractions.csproj">
```
Comment (bganapa) [inline src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj]: Remove commented project references

File: src/StackAdmin/Storage.Management/AzureRM.Storage.psd1 (hunk 2 above)
Comment (deathly809) [reply src/StackAdmin/Storage.Management/AzureRM.Storage.psd1]: Are we going to remove the dependency?

File: src/StackAdmin/Storage.Management/AzureRM.Storage.psd1 [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+CompanyName = 'Microsoft Corporation'
+
+# Copyright statement for this module
+Copyright = 'Microsoft Corporation. All rights reserved.'
+
+# Description of the functionality provided by this module
+Description = 'Microsoft Azure PowerShell - Storage service management cmdlets for Azure Resource Manager.  Creates and manages storage accounts in Azure Resource Manager.'
+
+# Minimum version of the Windows PowerShell engine required by this module
+PowerShellVersion = '3.0'
```
Comment (deathly809) [inline src/StackAdmin/Storage.Management/AzureRM.Storage.psd1]: Change to match AzureRM.Profile

File: src/StackAdmin/Storage.Management/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj [hunk 5]
```diff
@@ -40,10 +40,26 @@
     <Compile Include="TestController.cs" />
   </ItemGroup>
   <ItemGroup>
+    <!--    <ProjectReference Include="..\..\..\Common\Commands.Common.Authentication.Abstractions\Commands.Common.Authentication.Abstractions.csproj">
```
Comment (deathly809) [inline src/StackAdmin/Storage.Management/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj]: Remove commented out code please.

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj [hunk 6]
```diff
@@ -49,76 +49,55 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
+    <!-- We Moved this to a project we build: Microsoft.Azure.Management.Storage.6.4.0-preview -->
+    <Reference Include="Microsoft.Azure.KeyVault.Core">
+      <HintPath>..\..\..\..\packages\Microsoft.Azure.KeyVault.Core.1.0.0\lib\net40\Microsoft.Azure.KeyVault.Core.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <!--    <Reference Include="Microsoft.Azure.Management.Storage, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (deathly809) [inline src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj]: You could shorten this if you wanted the reference version.

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj [hunk 7]
```diff
@@ -49,76 +49,55 @@
...
       <Private>True</Private>
     </Reference>
     <Reference Include="System.Spatial">
-      <HintPath>..\..\..\packages\System.Spatial.5.6.4\lib\net40\System.Spatial.dll</HintPath>
+      <HintPath>..\..\..\packages\System.Spatial.5.8.2\lib\net40\System.Spatial.dll</HintPath>
       <Private>True</Private>
     </Reference>
   </ItemGroup>
   <ItemGroup>
-    <Compile Include="$(StorageToolsPath)Adapters\ARM.Storage.3\ARMStorageProvider.cs">
```
Comment (deathly809) [inline src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj]: You should use $(StorageToolsPath)

**Comment by bganapa** [comment]
@maddieclayton the PR is failing due to this test In Profile. I have already retried..

System.ArgumentException : An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.Azure.Commands.Profile.Test.TelemetryTests.DataCollectionHandlesSerializationErrors() in c:\workspace\powershell\src\ResourceManager\Profile\Commands.Profile.Test\TelemetryTests.cs:line

**Comment by MiYanni** [comment]
@bganapa We skipped `DataCollectionHandlesSerializationErrors` in the latest from the `preview` branch. Pull in to get that update.

File: src/StackAdmin/Storage.Management/AzureRM.Storage.psd1 (hunk 2 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/AzureRM.Storage.psd1]: I do not think we should remove it, if we remove, there may be a need to explicilty load the azure.storage module

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj (hunk 3 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj]: removed

File: src/StackAdmin/Storage.Management/AzureRM.Storage.psd1 (hunk 4 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/AzureRM.Storage.psd1]: done

File: src/StackAdmin/Storage.Management/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj (hunk 5 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj]: done

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj (hunk 6 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj]: done

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj (hunk 7 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj]: for some reason, storagetoolspath nuget package was failing , we should have got the package from common depepndencies.. let  me check again

**Comment by deathly809** [comment]
@maddieclayton  / @MiYanni  I am very confused on how the tests are passing as we don't have the Azure.Storage module in the repo anywhere.   I don't see any output about the tests actually running.

File: src/StackAdmin/Storage/Commands.Management.Storage/Commands.Management.Storage.csproj [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    <AssemblyOriginatorKeyFile>MSSharedLibKey.snk</AssemblyOriginatorKeyFile>
+    <DelaySign>true</DelaySign>
+    <Prefer32Bit>false</Prefer32Bit>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="Microsoft.Azure.KeyVault.Core">
+      <HintPath>..\..\..\..\packages\Microsoft.Azure.KeyVault.Core.1.0.0\lib\net40\Microsoft.Azure.KeyVault.Core.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <!--    <Reference Include="Microsoft.Azure.Management.Storage, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (deathly809) [inline src/StackAdmin/Storage/Commands.Management.Storage/Commands.Management.Storage.csproj]: Commented out code.

File: src/StackAdmin/Storage.Management/Commands.Management.Storage.Test/packages.config [hunk 9]
```diff
@@ -14,7 +14,7 @@
   <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="net45" />
   <package id="Microsoft.IdentityModel.Clients.ActiveDirectory" version="2.28.3" targetFramework="net45" />
   <package id="Microsoft.Net.Http" version="2.2.28" targetFramework="net45" />
-  <package id="Microsoft.Rest.ClientRuntime" version="2.3.12" targetFramework="net452" />
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.11" targetFramework="net452" />
```
Comment (markcowl) [inline src/StackAdmin/Storage.Management/Commands.Management.Storage.Test/packages.config]: This should not be downgraded

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/AzureRM.Storage.psd1 [hunk 10]
```diff
@@ -51,8 +51,8 @@ CLRVersion = '4.0'
 # ProcessorArchitecture = ''
 
 # Modules that must be imported into the global environment prior to importing this module
-RequiredModules = @(@{ModuleName = 'AzureRM.Profile'; ModuleVersion = '4.1.1'; }, 
-               @{ModuleName = 'Azure.Storage'; ModuleVersion = '1.0.5.4'; })
+RequiredModules = @(@{ModuleName = 'AzureRM.Profile'; ModuleVersion = '5.3.4'; },
```
Comment (markcowl) [inline src/StackAdmin/Storage.Management/Commands.Management.Storage/AzureRM.Storage.psd1]: This should match the current version

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj [hunk 11]
```diff
@@ -49,76 +49,51 @@
     <Prefer32Bit>false</Prefer32Bit>
   </PropertyGroup>
   <ItemGroup>
+    <!-- We Moved this to a project we build: Microsoft.Azure.Management.Storage.6.4.0-preview -->
```
Comment (markcowl) [inline src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj]: Use $LibrarySOurceFolder here

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Microsoft.Azure.Commands.Management.Storage.dll-Help.psd1 [hunk 12]
```diff
@@ -0,0 +1,111 @@
+#
```
Comment (markcowl) [inline src/StackAdmin/Storage.Management/Commands.Management.Storage/Microsoft.Azure.Commands.Management.Storage.dll-Help.psd1]: You don't need this

File: src/StackAdmin/Storage.Management/AzureRM.Storage.psd1 (hunk 1 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/AzureRM.Storage.psd1]: updated to 5.5.1

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj (hunk 7 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj]: changed

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/AzureRM.Storage.psd1 (hunk 10 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/Commands.Management.Storage/AzureRM.Storage.psd1]: updated to 5.5.1

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj (hunk 11 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/Commands.Management.Storage/Commands.Management.Storage.csproj]: changed to $(LibraryNugetPackageFolder)

**Comment by bganapa** [comment]
ps-sign - https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/ps-sign/1152/console
ps-demand - https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/1075/

**Comment by bganapa** [comment]
ps-netsign - https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/psnetsign/149/

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/Properties/AssemblyInfo.cs [hunk 13]
```diff
@@ -46,5 +46,5 @@
 // You can specify all the values or you can default the Build and Revision Numbers 
 // by using the '*' as shown below:
 
-[assembly: AssemblyVersion("3.3.2")]
-[assembly: AssemblyFileVersion("3.3.2")]
+[assembly: AssemblyVersion("3.3.3")]
```
Comment (maddieclayton) [inline src/StackAdmin/Storage.Management/Commands.Management.Storage/Properties/AssemblyInfo.cs]: Don't think this is the right version?

File: src/StackAdmin/Storage.Management/SDK/Properties/AssemblyInfo.cs [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+//
+//      Major Version
+//      Minor Version
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers
+// by using the '*' as shown below:
+
+[assembly: AssemblyVersion("1.0.0")]
```
Comment (maddieclayton) [inline src/StackAdmin/Storage.Management/SDK/Properties/AssemblyInfo.cs]: Need to be updated to the correct version.

File: src/StackAdmin/Storage.Management/SDK/StorageSDK.csproj [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    <DelaySign>true</DelaySign>
+    <Prefer32Bit>false</Prefer32Bit>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="Generated\**\*.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\..\packages\Newtonsoft.Json.6.0.8\lib\net45\Newtonsoft.Json.dll</HintPath>
```
Comment (maddieclayton) [inline src/StackAdmin/Storage.Management/SDK/StorageSDK.csproj]: $(LibraryNugetPackageFolder) - apply everywhere in this file.

File: src/StackAdmin/Storage/AzureRM.Storage.psd1 [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+# Generated on: 7/6/2017
+#
+
+@{
+
+# Script module or binary module file associated with this manifest.
+# RootModule = ''
+
+# Version number of this module.
+ModuleVersion = '1.1.0.1'
```
Comment (maddieclayton) [inline src/StackAdmin/Storage/AzureRM.Storage.psd1]: Correct version?

File: src/StackAdmin/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj [hunk 17]
```diff
@@ -0,0 +1,198 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\..\..\packages\xunit.runner.visualstudio.2.1.0\build\net20\xunit.runner.visualstudio.props" Condition="Exists('..\..\..\packages\xunit.runner.visualstudio.2.1.0\build\net20\xunit.runner.visualstudio.props')" />
```
Comment (maddieclayton) [inline src/StackAdmin/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj]: $(LibraryNugetPackageFolder) - apply everywhere in file.

File: src/StackAdmin/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+      <Private>True</Private>
+    </Reference>
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="ScenarioTests\StorageAccountTests.cs" />
+    <Compile Include="TestController.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\..\Common\Commands.Common.Authentication.Abstractions\Commands.Common.Authentication.Abstractions.csproj">
```
Comment (maddieclayton) [inline src/StackAdmin/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj]: Remove all common project references.

File: src/StackAdmin/Storage.Management/AzureRM.Storage.psd1 [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+# Generated on: 7/6/2017
+#
+
+@{
+
+# Script module or binary module file associated with this manifest.
+# RootModule = ''
+
+# Version number of this module.
+ModuleVersion = '1.1.0.1'
```
[the remark this answers is not in the document]
Comment (bganapa) [reply src/StackAdmin/Storage.Management/AzureRM.Storage.psd1]: Yes

[the remark this answers is not in the document]
Comment (bganapa) [reply src/StackAdmin/Storage.Management/AzureRM.Storage.psd1]: it is same as 1.1.0 except for common runtime changes and istoragecontext changes

File: src/StackAdmin/Storage.Management/Commands.Management.Storage/AzureRM.Storage.psd1 [hunk 20]
```diff
@@ -12,7 +12,7 @@
 # RootModule = ''
 
 # Version number of this module.
-ModuleVersion = '1.0.5.4'
+ModuleVersion = '1.1.0.1'
```
[the remark this answers is not in the document]
Comment (bganapa) [reply src/StackAdmin/Storage.Management/Commands.Management.Storage/AzureRM.Storage.psd1]: removed the one at the top level, it should not be there.. thanks for catching

File: src/StackAdmin/Storage.Management/SDK/StorageSDK.csproj (hunk 15 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/SDK/StorageSDK.csproj]: updated

File: src/StackAdmin/Storage.Management/SDK/Properties/AssemblyInfo.cs (hunk 14 above)
Comment (bganapa) [reply src/StackAdmin/Storage.Management/SDK/Properties/AssemblyInfo.cs]: updated to 6.4.1, 6.4.0 is the one we copied

**Comment by bganapa** [comment]
@maddieclayton addressed the review comments. Thanks for the review

**Comment by maddieclayton** [comment]
@bganapa Made two small changes but everything else looks good.  Please get @deathly809 to approve and we will merge tomorrow morning.

sign job is passing here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/ps-sign/1156/

**Comment by bganapa** [comment]
@deathly809 , @knithinc  to approve this PR

**Comment by deathly809** [comment]
@maddieclayton  I approved, but the build is failing trying to fetch from github.

**Comment by maddieclayton** [comment]
@deathly809 The build was previously failing to I pushed out a fix - we should see in about an hour if it fixed the build.

**Comment by bganapa** [comment]
@maddieclayton rebased to resolve conflicts in stack.sln

**Pull request merged**